pkgs/development/haskell-modules/make-package-set.nix

```diff
@@ -87,8 +87,11 @@ let
       drv = if lib.isFunction fn then fn else import fn;
       auto = builtins.intersectAttrs (lib.functionArgs drv) scope;
 
+      # Converts a returned function to a functor attribute set if necessary
+      ensureAttrs = v: if builtins.isFunction v then { __functor = _: v; } else v;
+
       # this wraps the `drv` function to add a `overrideScope` function to the result.
-      drvScope = allArgs: drv allArgs // {
+      drvScope = allArgs: ensureAttrs (drv allArgs) // {
```

Comment:
In what situation is it critical that `ensureAttrs` is used here?  (I'm mostly just curious...)

Reply:
When `drv allArgs` returns a function.

Reply:
Ah sorry, I can see that, but what I'm trying to ask is what sort of call to `callPackage` would fail if `ensureAttrs` was not used here?

For instance, I don't think `ensureAttrs` is necessary for a normal call to `callPackage` that looks like:

```nix
callPackage
  ( { mkDerivation, lens, aeson }:
    mkDerivation {
      name = "my-haskell-package";
      ...
    }
  )
  {}
```

What would a call to `callPackage` look like where `drv allArgs` returns function?  What would be the use of this?

Reply:
`callPackage ./hoogle.nix { }` and `callPackage ./with-packages-wrapper.nix { }` both need it.

Reply:
I guess we should have documented this feature, since it appears to differ from the top-level `callPackage`:

Here's an example showing what I mean:

```nix
with import <nixpkgs> {};

let
  myPkgExpression = { runCommand, lens }: prefix:
    runCommand "get-lens-version" {} ''
      echo ${prefix} lens version ${lens.version} >> $out
    '';

  myPkg = callPackage myPkg { lens = haskellPackages.lens; };

  myPkgWithArg = myPkg "my prefix";

  myPkgWithOverride = myPkg.override { lens = haskellPackages.lens_5_0_1; };

  myPkgWithArgWithOverride = myPkgWithArg.override { lens = haskellPackages.lens_5_0_1; };

  myPkgWithOverrideWithArg = myPkgWithOverride "my prefix";
in

myPkgWithArgWithOverride
# myPkgWithOverrideWithArg
```

`myPkgExpression` (and myPkg) define a package using `callPackage` similar to our new `hoogle.nix` and `with-packages-wrapper.nix`.

`myPkgWithOverride` and `myPkgWithArgWithOverride` then try to override the `lens` version, both before and after passing the actual `"my prefix"` argument.

With both of these, I get the following error:

```
error: infinite recursion encountered

       at /nix/store/g65b6lh3akdka0z9iji75yl07sb9n2s6-nixos-21.11.334247.573095944e7/nixos/lib/trivial.nix:364:19:

          363|   */
          364|   isFunction = f: builtins.isFunction f ||
             |                   ^
          365|     (f ? __functor && isFunction (f.__functor f));
```

I guess we should have explicitly documented this in `haskellPackages.callPackage` so that users see that `haskellPackages.callPackage` is a little more flexible than the top-level `callPackage`.

Reply:
I'd rather say that this is a potential bug in top level `callPackage` or a pitfall — in principle the behavior of top level `callPackage` is the same: cnovert a function to a functor set in order to be able to add the `override` attribute.